app/javascript/extensible-components/index.ts

```diff
@@ -0,0 +1,67 @@
+import { IExtensionComponent, IMiQApiCallback } from './lib';
+
+/**
+ * Class for easy creation of extensible component.
+ */
+export class ExtensibleComponent {
+  public unsubscribe: Function;
+  constructor(public name: string, public api: IMiQApiCallback, public render: IMiQApiCallback){}
+}
+
+/**
+ * Create new object which will hold extension components on MiQ main object.
+ */
+ManageIQ.extensionComponents = ManageIQ.extensionComponents || {};
+
+/**
+ * Subject from from Rxjs to send message that we want to register new component.
+ */
+ManageIQ.extensionComponents.source = new Rx.Subject();
+
+/**
+ * Components will be saved in items which will be set. To easy remove of components which no longer exists.
+ */
+ManageIQ.extensionComponents.items = new Set();
+
+/**
+ * Helper function to create new component.
+ * @param name string name of new component.
+ * @param api callback functions to change inner logic of component.
+ * @param render callback function to apply render functions.
```

Comment:
Looking at [MiQ plugin example using this PR](https://github.com/martinpovolny/miq_plugin_example/pull/5/files#diff-4afee5c4ac902dbdd9e098ec151c0892), I assume the render object serves the purpose of visual (DOM level) integration, correct?

However, I'd really suggest using the concept of render functions instead of DOM element getters.

A render function is a function which takes a DOM element and renders something into it:

```
function renderStuff(element) { ... }
```

The `render` parameter here should be an object containing render functions, using following interface signature:

```
[propName: string]: (element: HTMLElement) => void;
```

Example code using a render function:

```
// component is obtained through subscribe API
component.render.addNewButton((element) => {
  // actual render logic goes here, e.g. ReactDOM.render
});
```

Reply:
I'd put both `getPlaceToRender` and standalone `render` function at same level (render callbacks). I guess naming this render is wrong, how about `domCallbacks`? Generally we want to enable both render from component itself and get element to render inside plugin.

Reply:
Well, the `render` object was meant as a hook to the DOM tree owned by the given component. This object would contain render callbacks - functions of type `(element: HTMLElement) => void`.

```
tableComponent.render.actionButton((element) => ...)
tableComponent.render.customFooter((element) => ...)
```

The `element` argument represents the DOM container to be used for extending the component's own DOM tree. The `element` is owned by the component itself, acting as a bridge between element's own DOM and the custom extension's DOM.

For example, imagine an Angular component with following DOM:

```
<div class='miq-table'> <!-- component root element -->
  <div class='miq-table-buttons'>
    <button>one</button> <button>two</button>
    <span /> <!-- bridge for render.actionButton -->
  </div>
  <div class='miq-table-content />
  <div class='miq-table-footer>
    <span>foo text</span>
    <span /> <!-- bridge for render.customFooter -->
  </div>
</div>
```

The functions are designed as callbacks, which gives the component control over when they are called, because the component itself knows best when to call them, and if to call them at all.

If we expose the component's DOM element (e.g. `getPlaceToRender`), we risk the chance that component's own DOM tree gets out of sync with given DOM element (imagine a case where a table component would hide its footer). I think it's better to just use the render callbacks with meaningful names & giving control over their execution to the component itself. @karelhala what do you think?

Reply:
For simple html changes, it's good. However if some plugin would like to use their own render function (for example react) we'd loose this option.

I'd allow both of these functions:
1) simple HTML render like creating new input which can be either some HTML partial or object with name, ID, etc..
2) for complex rendering I'd allow return of placeholder element. In this callback function we can have some logic which will check if it was called (so some other plugin won't break anything).

Developer of such plugin has to know that using such functions can break something and it's his responsibility to change something. However with second function we have a lot of advantages so we should consider if we really don't want to use it. For me it feels less invasive to pass placeholder element and let plugin render some stuff inside it, rather than doing something like this:
```javascript
let placeholder = document.createElement('div');
ReactDOM.render(<someCmp />, placeholder);

var subs = subscribe('toolbar');
subs.with((component) => component.renderSomeContent(placeholder));
```
Where `renderSomeContent` is function for rendering HTML partial inside extended component. I don't believe it will behave correctly and that data changes will be refreshed correctly.

However I understand that the second option for complex examples should not be part of render object, rather to be part of API object.

So the question here is. Do we want to render just plain HTML partials or let plugins render it themselves via their own logic? I'd go for both options.